prov/sm2/src/sm2_progress.c

```diff
@@ -42,52 +42,187 @@
 #include "sm2.h"
 #include "sm2_fifo.h"
 
-static inline int sm2_cma_loop(pid_t pid, struct iovec *local,
-			       unsigned long local_cnt, struct iovec *remote,
-			       unsigned long remote_cnt, size_t total,
-			       bool write)
+static int sm2_progress_ipc_host_to_dev(struct sm2_ep *ep,
+					struct sm2_xfer_entry *xfer_entry,
+					struct fi_peer_rx_entry *rx_entry,
+					void *host_ptr)
 {
+	sm2_gid_t sender_gid = xfer_entry->hdr.sender_gid;
+	struct sm2_ipc_host_to_dev_data *host_to_dev_data;
+	struct ipc_info *ipc_info;
+	uint64_t device;
+	void *device_ptr = rx_entry->iov[0].iov_base;
+	struct ofi_mr **mr = (struct ofi_mr **) rx_entry->desc;
+	enum fi_hmem_iface iface = mr[0]->iface;
+	void *base;
+	int ret;
+
+	if (hmem_ops[iface].is_addr_valid(device_ptr, &device, NULL)) {
+		/* The destination address is a valid device address */
+		xfer_entry->hdr.proto = sm2_proto_ipc_host_to_dev;
+		xfer_entry->hdr.sender_gid = ep->gid;
+
+		host_to_dev_data = ((struct sm2_ipc_host_to_dev_data *)
+					    xfer_entry->user_data);
+		host_to_dev_data->host_ptr = host_ptr;
+		host_to_dev_data->rx_context = rx_entry->context;
+		host_to_dev_data->rx_flags = rx_entry->flags;
+
+		ipc_info = &host_to_dev_data->ipc_info;
+		ipc_info->iface = iface;
+		ipc_info->device = device;
+
+		/* Open IPC handle */
+		ret = ofi_hmem_get_base_addr(ipc_info->iface, device_ptr, &base,
+					     &ipc_info->base_length);
+		if (ret) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed to get device memory base address. "
+				"Error code: %d\n",
+				ret);
+			return ret;
+		}
+
+		ret = ofi_hmem_get_handle(ipc_info->iface, base,
+					  ipc_info->base_length,
+					  (void **) &ipc_info->ipc_handle);
+		if (ret) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed to open IPC handle in host to device "
+				"protocol. Error code: %d\n",
+				ret);
+			return ret;
+		}
+
+		ipc_info->base_addr = (uintptr_t) base;
+		ipc_info->offset = (uintptr_t) device_ptr -
+				   (uintptr_t) ipc_info->base_addr;
+
+		/* Send xfer_entry with IPC handle back to sender */
+		sm2_fifo_write(ep, sender_gid, xfer_entry);
+
+		return FI_SUCCESS;
+	} else {
+		assert(1 == 0);
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL, "CMA error %d\n", errno);
+		return -FI_EIO;
+	}
+}
+
+static int sm2_ipc_host_to_dev_hmem_memcpy(struct sm2_ep *ep,
```

Comment:
Can we reuse/consolidate this with the normal ipc path? Same with the get base addr/open path. Also some of these names are getting long. Consolidating would help simplify them/make them more generic.